File: leetcode/problem221.py
import math

class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        memo = {}
        h = len(matrix)
        if h == 0:
            return 0
        w = len(matrix[0])


        answer = 0
        for y in range(h+1):
            for x in range(w+1):
                if y == 0 or x == 0 or matrix[y-1][x-1] == '0':
                    memo[(y, x)] = 0
                    continue

                surround_min = min(memo[(y-1, x-1)], memo[(y-1,x)], memo[(y, x-1)])
                if surround_min != 0:
                    sqrt_min = math.sqrt(surround_min)
                    memo[(y, x)] = pow(sqrt_min + 1, 2)
                else:
                    memo[(y, x)] = 1

                answer = max(answer, memo[(y, x)])

        return int(answer)

File: leetcode/problem221.cpp
class Solution {
public:
int maximalSquare(vector<vector<char>> matrix) {
    int N = matrix.size();

    if(N == 0) {
        return 0;
    }

    int M = matrix[0].size();

    vector<vector<int>> dp(N, vector(M, 0));

    int answer = 1;
    bool is_zero = true;

    for (int i=0; i<N; ++i) {
        for(int j=0; j<M; ++j) {
            if(matrix[i][j] == '1') {
                is_zero = false;
                dp[i][j] = 1;

                if(i-1 >= 0 && j-1 >= 0) {
                    int min_len = min(dp[i-1][j], min(dp[i][j-1], dp[i-1][j-1]));
                    if(min_len) {
                        dp[i][j] = min_len + 1;
                        answer = max(answer, min_len + 1);
                    }
                }
            }
        }
    }
    return is_zero ? 0 : answer * answer;
}
};

